[SPARK-21786][SQL] The 'spark.sql.parquet.compression.codec' and 'spark.sql.orc.compression.codec' configuration doesn't take effect on hive table writing

fjh100456 opened the pull request:
[SPARK-21786][SQL] The 'spark.sql.parquet.compression.codec' and 'spark.sql.orc.compression.codec' configuration doesn't take effect on hive table writing

What changes were proposed in this pull request?

Pass ‘spark.sql.parquet.compression.codec’ value to ‘parquet.compression’.
Pass ‘spark.sql.orc.compression.codec’ value to ‘orc.compress’.

How was this patch tested?

Add test.

Note: 
This is the same issue mentioned in #19218 . That branch was deleted mistakenly, so make a new pr instead.

@gatorsmile @maropu @dongjoon-hyun @discipleforteen

fjh100456 committed:
[SPARK-21786][SQL] When acquiring 'compressionCodecClassName' in 'ParquetOptions', `parquet.compression` needs to be considered.

## What changes were proposed in this pull request?
1.Increased acquiring 'compressionCodecClassName' from `parquet.compression`,and the order is `compression`,`parquet.compression`,`spark.sql.parquet.compression.codec`, just like what we do in `OrcOptions`.
2.Change `spark.sql.parquet.compression.codec` to support "none".Actually in `ParquetOptions`,we do support "none" as equivalent to "uncompressed", but it does not allowed to configured to "none".

## How was this patch tested?
Manual test.

fjh100456 committed:
[SPARK-21786][SQL] When acquiring 'compressionCodecClassName' in 'ParquetOptions', `parquet.compression` needs to be considered.

## What changes were proposed in this pull request?
1.Increased acquiring 'compressionCodecClassName' from `parquet.compression`,and the order is `compression`,`parquet.compression`,`spark.sql.parquet.compression.codec`, just like what we do in `OrcOptions`.
2.Change `spark.sql.parquet.compression.codec` to support "none".Actually in `ParquetOptions`,we do support "none" as equivalent to "uncompressed", but it does not allowed to configured to "none".

## How was this patch tested?
Manual test.

fjh100456 committed:
spark.sql.parquet.compression.codec[SPARK-21786][SQL] When acquiring 'compressionCodecClassName' in 'ParquetOptions', `parquet.compression` needs to be considered.

## What changes were proposed in this pull request?
1.Increased acquiring 'compressionCodecClassName' from `parquet.compression`,and the order is `compression`,`parquet.compression`,`spark.sql.parquet.compression.codec`, just like what we do in `OrcOptions`.
2.Change `spark.sql.parquet.compression.codec` to support "none".Actually in `ParquetOptions`,we do support "none" as equivalent to "uncompressed", but it does not allowed to configured to "none".

## How was this patch tested?
Manual test.

fjh100456 committed:
spark.sql.parquet.compression.codec[SPARK-21786][SQL] When acquiring 'compressionCodecClassName' in 'ParquetOptions', `parquet.compression` needs to be considered.

## What changes were proposed in this pull request?
1.Increased acquiring 'compressionCodecClassName' from `parquet.compression`,and the order is `compression`,`parquet.compression`,`spark.sql.parquet.compression.codec`, just like what we do in `OrcOptions`.
2.Change `spark.sql.parquet.compression.codec` to support "none".Actually in `ParquetOptions`,we do support "none" as equivalent to "uncompressed", but it does not allowed to configured to "none".
3.Change `compressionCode` to `compressionCodecClassName`.

## How was this patch tested?
Manual test.

fjh100456 committed:
Make comression codec take effect in hive table writing.

fjh100456 committed:
Modify test

fjh100456 committed:
Separate the pr

fjh100456 committed:
Add test case with the table containing mixed compression codec

fjh100456 committed:
Revert back

fjh100456 committed:
Revert back

fjh100456 committed:
Add a new line at the of file

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala:
```diff
@@ -68,6 +68,10 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
 
+    // Set compression by priority
+    HiveOptions.getHiveWriteCompression(fileSinkConf.getTableInfo, sparkSession.sessionState.conf)
+      .foreach { case (compression, codec) => hadoopConf.set(compression, codec) }
```
This will not be affected by `hive.exec.compress.output`? Could you do an investigation the relation of this setting with the codes from line 57 to line 69?

fjh100456 replied:
For parquet, without the changes of this pr, the precedence is table-level compression > `mapreduce.output.fileoutputformat.compress`. `spark.sql.parquet.compression` never takes effect. But now with this pr, `mapreduce.output.fileoutputformat.compress` will not take effect. As an alternative, `spark.sql.parquet.compression` will always take effect if there is no table level compression.

For ORC, `hive.exec.compress.output` does not take effect, as explained in the comments of the Code.

Shall we keep this precedence for parquet? If so, how to deal with ORC?

SparkQA commented:
**[Test build #85539 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85539/testReport)** for PR 20087 at commit [`ee0c558`](https://github.com/apache/spark/commit/ee0c5587374ed33e29a763f8764263114a9b57ab).

SparkQA commented:
**[Test build #85539 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85539/testReport)** for PR 20087 at commit [`ee0c558`](https://github.com/apache/spark/commit/ee0c5587374ed33e29a763f8764263114a9b57ab).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85539/
Test FAILed.

gatorsmile committed:
[SPARK-22932][SQL] Refactor AnalysisContext

## What changes were proposed in this pull request?
Add a `reset` function to ensure the state in `AnalysisContext ` is per-query.

## How was this patch tested?
The existing test cases

Author: gatorsmile <gatorsmile@gmail.com>

Closes #20127 from gatorsmile/refactorAnalysisContext.

fjh100456 committed:
Fix scala style

SparkQA commented:
**[Test build #85582 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85582/testReport)** for PR 20087 at commit [`e9f705d`](https://github.com/apache/spark/commit/e9f705d0ad783da5bd091632e98a6151d4d21cb6).

SparkQA commented:
**[Test build #85582 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85582/testReport)** for PR 20087 at commit [`e9f705d`](https://github.com/apache/spark/commit/e9f705d0ad783da5bd091632e98a6151d4d21cb6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85582/
Test FAILed.

fjh100456 committed:
Fix scala style

SparkQA commented:
**[Test build #85583 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85583/testReport)** for PR 20087 at commit [`d3aa7a0`](https://github.com/apache/spark/commit/d3aa7a01320b6af2866d7cf7c4f178eb23eae3ad).

SparkQA commented:
**[Test build #85583 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85583/testReport)** for PR 20087 at commit [`d3aa7a0`](https://github.com/apache/spark/commit/d3aa7a01320b6af2866d7cf7c4f178eb23eae3ad).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85583/
Test FAILed.

advancedxy committed:
[SPARK-22897][CORE] Expose stageAttemptId in TaskContext

## What changes were proposed in this pull request?
stageAttemptId added in TaskContext and corresponding construction modification

## How was this patch tested?
Added a new test in TaskContextSuite, two cases are tested:
1. Normal case without failure
2. Exception case with resubmitted stages

Link to [SPARK-22897](https://issues.apache.org/jira/browse/SPARK-22897)

Author: Xianjin YE <advancedxy@gmail.com>

Closes #20082 from advancedxy/SPARK-22897.

(cherry picked from commit a6fc300e91273230e7134ac6db95ccb4436c6f8f)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

juliuszsompolski committed:
[SPARK-22938] Assert that SQLConf.get is accessed only on the driver.

## What changes were proposed in this pull request?

Assert if code tries to access SQLConf.get on executor.
This can lead to hard to detect bugs, where the executor will read fallbackConf, falling back to default config values, ignoring potentially changed non-default configs.
If a config is to be passed to executor code, it needs to be read on the driver, and passed explicitly.

## How was this patch tested?

Check in existing tests.

Author: Juliusz Sompolski <julek@databricks.com>

Closes #20136 from juliuszsompolski/SPARK-22938.

(cherry picked from commit 247a08939d58405aef39b2a4e7773aa45474ad12)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

gatorsmile committed:
[SPARK-22934][SQL] Make optional clauses order insensitive for CREATE TABLE SQL statement

## What changes were proposed in this pull request?
Currently, our CREATE TABLE syntax require the EXACT order of clauses. It is pretty hard to remember the exact order. Thus, this PR is to make optional clauses order insensitive for `CREATE TABLE` SQL statement.

```
CREATE [TEMPORARY] TABLE [IF NOT EXISTS] [db_name.]table_name
    [(col_name1 col_type1 [COMMENT col_comment1], ...)]
    USING datasource
    [OPTIONS (key1=val1, key2=val2, ...)]
    [PARTITIONED BY (col_name1, col_name2, ...)]
    [CLUSTERED BY (col_name3, col_name4, ...) INTO num_buckets BUCKETS]
    [LOCATION path]
    [COMMENT table_comment]
    [TBLPROPERTIES (key1=val1, key2=val2, ...)]
    [AS select_statement]
```

The proposal is to make the following clauses order insensitive.
```
    [OPTIONS (key1=val1, key2=val2, ...)]
    [PARTITIONED BY (col_name1, col_name2, ...)]
    [CLUSTERED BY (col_name3, col_name4, ...) INTO num_buckets BUCKETS]
    [LOCATION path]
    [COMMENT table_comment]
    [TBLPROPERTIES (key1=val1, key2=val2, ...)]
```

The same idea is also applicable to Create Hive Table.
```
CREATE [EXTERNAL] TABLE [IF NOT EXISTS] [db_name.]table_name
    [(col_name1[:] col_type1 [COMMENT col_comment1], ...)]
    [COMMENT table_comment]
    [PARTITIONED BY (col_name2[:] col_type2 [COMMENT col_comment2], ...)]
    [ROW FORMAT row_format]
    [STORED AS file_format]
    [LOCATION path]
    [TBLPROPERTIES (key1=val1, key2=val2, ...)]
    [AS select_statement]
```

The proposal is to make the following clauses order insensitive.
```
    [COMMENT table_comment]
    [PARTITIONED BY (col_name2[:] col_type2 [COMMENT col_comment2], ...)]
    [ROW FORMAT row_format]
    [STORED AS file_format]
    [LOCATION path]
    [TBLPROPERTIES (key1=val1, key2=val2, ...)]
```

## How was this patch tested?
Added test cases

Author: gatorsmile <gatorsmile@gmail.com>

Closes #20133 from gatorsmile/createDataSourceTableDDL.

(cherry picked from commit 1a87a1609c4d2c9027a2cf669ea3337b89f61fb6)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

cloud-fan committed:
[SPARK-20236][SQL] dynamic partition overwrite

## What changes were proposed in this pull request?

When overwriting a partitioned table with dynamic partition columns, the behavior is different between data source and hive tables.

data source table: delete all partition directories that match the static partition values provided in the insert statement.

hive table: only delete partition directories which have data written into it

This PR adds a new config to make users be able to choose hive's behavior.

## How was this patch tested?

new tests

Author: Wenchen Fan <wenchen@databricks.com>

Closes #18714 from cloud-fan/overwrite-partition.

(cherry picked from commit a66fe36cee9363b01ee70e469f1c968f633c5713)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

chetkhatri committed:
[SPARK-22896] Improvement in String interpolation

## What changes were proposed in this pull request?

* String interpolation in ml pipeline example has been corrected as per scala standard.

## How was this patch tested?
* manually tested.

Author: chetkhatri <ckhatrimanjal@gmail.com>

Closes #20070 from chetkhatri/mllib-chetan-contrib.

(cherry picked from commit 9a2b65a3c0c36316aae0a53aa0f61c5044c2ceff)
Signed-off-by: Sean Owen <sowen@cloudera.com>

cloud-fan committed:
[SPARK-20960][SQL] make ColumnVector public

## What changes were proposed in this pull request?

move `ColumnVector` and related classes to `org.apache.spark.sql.vectorized`, and improve the document.

## How was this patch tested?

existing tests.

Author: Wenchen Fan <wenchen@databricks.com>

Closes #20116 from cloud-fan/column-vector.

(cherry picked from commit b297029130735316e1ac1144dee44761a12bfba7)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

cloud-fan committed:
[SPARK-22944][SQL] improve FoldablePropagation

## What changes were proposed in this pull request?

`FoldablePropagation` is a little tricky as it needs to handle attributes that are miss-derived from children, e.g. outer join outputs. This rule does a kind of stop-able tree transform, to skip to apply this rule when hit a node which may have miss-derived attributes.

Logically we should be able to apply this rule above the unsupported nodes, by just treating the unsupported nodes as leaf nodes. This PR improves this rule to not stop the tree transformation, but reduce the foldable expressions that we want to propagate.

## How was this patch tested?

existing tests

Author: Wenchen Fan <wenchen@databricks.com>

Closes #20139 from cloud-fan/foldable.

(cherry picked from commit 7d045c5f00e2c7c67011830e2169a4e130c3ace8)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

felixcheung committed:
[SPARK-22933][SPARKR] R Structured Streaming API for withWatermark, trigger, partitionBy

## What changes were proposed in this pull request?

R Structured Streaming API for withWatermark, trigger, partitionBy

## How was this patch tested?

manual, unit tests

Author: Felix Cheung <felixcheung_m@hotmail.com>

Closes #20129 from felixcheung/rwater.

(cherry picked from commit df95a908baf78800556636a76d58bba9b3dd943f)
Signed-off-by: Felix Cheung <felixcheung@apache.org>

yaooqinn committed:
[SPARK-22950][SQL] Handle ChildFirstURLClassLoader's parent

## What changes were proposed in this pull request?

ChildFirstClassLoader's parent is set to null, so we can't get jars from its parent. This will cause ClassNotFoundException during HiveClient initialization with builtin hive jars, where we may should use spark context loader instead.

## How was this patch tested?

add new ut
cc cloud-fan gatorsmile

Author: Kent Yao <yaooqinn@hotmail.com>

Closes #20145 from yaooqinn/SPARK-22950.

(cherry picked from commit 9fa703e89318922393bae03c0db4575f4f4b4c56)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

cloud-fan committed:
[SPARK-22945][SQL] add java UDF APIs in the functions object

## What changes were proposed in this pull request?

Currently Scala users can use UDF like
```
val foo = udf((i: Int) => Math.random() + i).asNondeterministic
df.select(foo('a))
```
Python users can also do it with similar APIs. However Java users can't do it, we should add Java UDF APIs in the functions object.

## How was this patch tested?

new tests

Author: Wenchen Fan <wenchen@databricks.com>

Closes #20141 from cloud-fan/udf.

(cherry picked from commit d5861aba9d80ca15ad3f22793b79822e470d6913)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

gatorsmile committed:
[SPARK-22939][PYSPARK] Support Spark UDF in registerFunction

## What changes were proposed in this pull request?
```Python
import random
from pyspark.sql.functions import udf
from pyspark.sql.types import IntegerType, StringType
random_udf = udf(lambda: int(random.random() * 100), IntegerType()).asNondeterministic()
spark.catalog.registerFunction("random_udf", random_udf, StringType())
spark.sql("SELECT random_udf()").collect()
```

We will get the following error.
```
Py4JError: An error occurred while calling o29.__getnewargs__. Trace:
py4j.Py4JException: Method __getnewargs__([]) does not exist
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:318)
	at py4j.reflection.ReflectionEngine.getMethod(ReflectionEngine.java:326)
	at py4j.Gateway.invoke(Gateway.java:274)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:745)
```

This PR is to support it.

## How was this patch tested?
WIP

Author: gatorsmile <gatorsmile@gmail.com>

Closes #20137 from gatorsmile/registerFunction.

(cherry picked from commit 5aadbc929cb194e06dbd3bab054a161569289af5)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

maropu committed:
[SPARK-22771][SQL] Add a missing return statement in Concat.checkInputDataTypes

## What changes were proposed in this pull request?
This pr is a follow-up to fix a bug left in #19977.

## How was this patch tested?
Added tests in `StringExpressionsSuite`.

Author: Takeshi Yamamuro <yamamuro@apache.org>

Closes #20149 from maropu/SPARK-22771-FOLLOWUP.

(cherry picked from commit 6f68316e98fad72b171df422566e1fc9a7bbfcde)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

jerryshao committed:
[SPARK-21475][CORE][2ND ATTEMPT] Change to use NIO's Files API for external shuffle service

## What changes were proposed in this pull request?

This PR is the second attempt of #18684 , NIO's Files API doesn't override `skip` method for `InputStream`, so it will bring in performance issue (mentioned in #20119). But using `FileInputStream`/`FileOutputStream` will also bring in memory issue (https://dzone.com/articles/fileinputstream-fileoutputstream-considered-harmful), which is severe for long running external shuffle service. So here in this proposal, only fixing the external shuffle service related code.

## How was this patch tested?

Existing tests.

Author: jerryshao <sshao@hortonworks.com>

Closes #20144 from jerryshao/SPARK-21475-v2.

(cherry picked from commit 93f92c0ed7442a4382e97254307309977ff676f8)
Signed-off-by: Shixiong Zhu <zsxwing@gmail.com>

Commit:
[SPARK-22850][CORE] Ensure queued events are delivered to all event queues.

The code in LiveListenerBus was queueing events before start in the
queues themselves; so in situations like the following:

   bus.post(someEvent)
   bus.addToEventLogQueue(listener)
   bus.start()

"someEvent" would not be delivered to "listener" if that was the first
listener in the queue, because the queue wouldn't exist when the
event was posted.

This change buffers the events before starting the bus in the bus itself,
so that they can be delivered to all registered queues when the bus is
started.

Also tweaked the unit tests to cover the behavior above.

Author: Marcelo Vanzin <vanzin@cloudera.com>

Closes #20039 from vanzin/SPARK-22850.

(cherry picked from commit d2cddc88eac32f26b18ec26bb59e85c6f09a8c88)
Signed-off-by: Imran Rashid <irashid@cloudera.com>

Commit:
[SPARK-22948][K8S] Move SparkPodInitContainer to correct package.

Author: Marcelo Vanzin <vanzin@cloudera.com>

Closes #20156 from vanzin/SPARK-22948.

(cherry picked from commit 95f9659abe8845f9f3f42fd7ababd79e55c52489)
Signed-off-by: Marcelo Vanzin <vanzin@cloudera.com>

liyinan926 committed:
[SPARK-22953][K8S] Avoids adding duplicated secret volumes when init-container is used

## What changes were proposed in this pull request?

User-specified secrets are mounted into both the main container and init-container (when it is used) in a Spark driver/executor pod, using the `MountSecretsBootstrap`. Because `MountSecretsBootstrap` always adds new secret volumes for the secrets to the pod, the same secret volumes get added twice, one when mounting the secrets to the main container, and the other when mounting the secrets to the init-container. This PR fixes the issue by separating `MountSecretsBootstrap.mountSecrets` out into two methods: `addSecretVolumes` for adding secret volumes to a pod and `mountSecrets` for mounting secret volumes to a container, respectively. `addSecretVolumes` is only called once for each pod, whereas `mountSecrets` is called individually for the main container and the init-container (if it is used).

Ref: https://github.com/apache-spark-on-k8s/spark/issues/594.

## How was this patch tested?
Unit tested and manually tested.

vanzin This replaces https://github.com/apache/spark/pull/20148.
hex108 foxish kimoonkim

Author: Yinan Li <liyinan926@gmail.com>

Closes #20159 from liyinan926/master.

(cherry picked from commit e288fc87a027ec1e1a21401d1f151df20dbfecf3)
Signed-off-by: Marcelo Vanzin <vanzin@cloudera.com>

Commit:
[SPARK-22960][K8S] Make build-push-docker-images.sh more dev-friendly.

- Make it possible to build images from a git clone.
- Make it easy to use minikube to test things.

Also fixed what seemed like a bug: the base image wasn't getting the tag
provided in the command line. Adding the tag allows users to use multiple
Spark builds in the same kubernetes cluster.

Tested by deploying images on minikube and running spark-submit from a dev
environment; also by building the images with different tags and verifying
"docker images" in minikube.

Author: Marcelo Vanzin <vanzin@cloudera.com>

Closes #20154 from vanzin/SPARK-22960.

(cherry picked from commit 0428368c2c5e135f99f62be20877bbbda43be310)
Signed-off-by: Marcelo Vanzin <vanzin@cloudera.com>

juliuszsompolski committed:
[SPARK-22957] ApproxQuantile breaks if the number of rows exceeds MaxInt

## What changes were proposed in this pull request?

32bit Int was used for row rank.
That overflowed in a dataframe with more than 2B rows.

## How was this patch tested?

Added test, but ignored, as it takes 4 minutes.

Author: Juliusz Sompolski <julek@databricks.com>

Closes #20152 from juliuszsompolski/SPARK-22957.

(cherry picked from commit df7fc3ef3899cadd252d2837092bebe3442d6523)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

maropu committed:
[SPARK-22825][SQL] Fix incorrect results of Casting Array to String

## What changes were proposed in this pull request?
This pr fixed the issue when casting arrays into strings;
```
scala> val df = spark.range(10).select('id.cast("integer")).agg(collect_list('id).as('ids))
scala> df.write.saveAsTable("t")
scala> sql("SELECT cast(ids as String) FROM t").show(false)
+------------------------------------------------------------------+
|ids                                                               |
+------------------------------------------------------------------+
|org.apache.spark.sql.catalyst.expressions.UnsafeArrayData8bc285df|
+------------------------------------------------------------------+
```

This pr modified the result into;
```
+------------------------------+
|ids                           |
+------------------------------+
|[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]|
+------------------------------+
```

## How was this patch tested?
Added tests in `CastSuite` and `SQLQuerySuite`.

Author: Takeshi Yamamuro <yamamuro@apache.org>

Closes #20024 from maropu/SPARK-22825.

(cherry picked from commit 52fc5c17d9d784b846149771b398e741621c0b5c)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

MrBago committed:
[SPARK-22949][ML] Apply CrossValidator approach to Driver/Distributed memory tradeoff for TrainValidationSplit

## What changes were proposed in this pull request?

Avoid holding all models in memory for `TrainValidationSplit`.

## How was this patch tested?

Existing tests.

Author: Bago Amirbekian <bago@databricks.com>

Closes #20143 from MrBago/trainValidMemoryFix.

(cherry picked from commit cf0aa65576acbe0209c67f04c029058fd73555c1)
Signed-off-by: Joseph K. Bradley <joseph@databricks.com>

liyinan926 committed:
[SPARK-22757][K8S] Enable spark.jars and spark.files in KUBERNETES mode

## What changes were proposed in this pull request?

We missed enabling `spark.files` and `spark.jars` in https://github.com/apache/spark/pull/19954. The result is that remote dependencies specified through `spark.files` or `spark.jars` are not included in the list of remote dependencies to be downloaded by the init-container. This PR fixes it.

## How was this patch tested?

Manual tests.

vanzin This replaces https://github.com/apache/spark/pull/20157.

foxish

Author: Yinan Li <liyinan926@gmail.com>

Closes #20160 from liyinan926/SPARK-22757.

(cherry picked from commit 6cff7d19f6a905fe425bd6892fe7ca014c0e696b)
Signed-off-by: Felix Cheung <felixcheung@apache.org>

adrian-ionescu committed:
[SPARK-22961][REGRESSION] Constant columns should generate QueryPlanConstraints

## What changes were proposed in this pull request?

#19201 introduced the following regression: given something like `df.withColumn("c", lit(2))`, we're no longer picking up `c === 2` as a constraint and infer filters from it when joins are involved, which may lead to noticeable performance degradation.

This patch re-enables this optimization by picking up Aliases of Literals in Projection lists as constraints and making sure they're not treated as aliased columns.

## How was this patch tested?

Unit test was added.

Author: Adrian Ionescu <adrian@databricks.com>

Closes #20155 from adrian-ionescu/constant_constraints.

(cherry picked from commit 51c33bd0d402af9e0284c6cbc0111f926446bfba)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

bersprockets committed:
[SPARK-22940][SQL] HiveExternalCatalogVersionsSuite should succeed on platforms that don't have wget

## What changes were proposed in this pull request?

Modified HiveExternalCatalogVersionsSuite.scala to use Utils.doFetchFile to download different versions of Spark binaries rather than launching wget as an external process.

On platforms that don't have wget installed, this suite fails with an error.

cloud-fan : would you like to check this change?

## How was this patch tested?

1) test-only of HiveExternalCatalogVersionsSuite on several platforms. Tested bad mirror, read timeout, and redirects.
2) ./dev/run-tests

Author: Bruce Robbins <bersprockets@gmail.com>

Closes #20147 from bersprockets/SPARK-22940-alt.

(cherry picked from commit c0b7424ecacb56d3e7a18acc11ba3d5e7be57c43)
Signed-off-by: Marcelo Vanzin <vanzin@cloudera.com>

jkbradley committed:
[SPARK-13030][ML] Follow-up cleanups for OneHotEncoderEstimator

## What changes were proposed in this pull request?

Follow-up cleanups for the OneHotEncoderEstimator PR.  See some discussion in the original PR: https://github.com/apache/spark/pull/19527 or read below for what this PR includes:
* configedCategorySize: I reverted this to return an Array.  I realized the original setup (which I had recommended in the original PR) caused the whole model to be serialized in the UDF.
* encoder: I reorganized the logic to show what I meant in the comment in the previous PR.  I think it's simpler but am open to suggestions.

I also made some small style cleanups based on IntelliJ warnings.

## How was this patch tested?

Existing unit tests

Author: Joseph K. Bradley <joseph@databricks.com>

Closes #20132 from jkbradley/viirya-SPARK-13030.

(cherry picked from commit 930b90a84871e2504b57ed50efa7b8bb52d3ba44)
Signed-off-by: Joseph K. Bradley <joseph@databricks.com>

gerashegalov committed:
[SPARK-22914][DEPLOY] Register history.ui.port

## What changes were proposed in this pull request?

Register spark.history.ui.port as a known spark conf to be used in substitution expressions even if it's not set explicitly.

## How was this patch tested?

Added unit test to demonstrate the issue

Author: Gera Shegalov <gera@apache.org>
Author: Gera Shegalov <gshegalov@salesforce.com>

Closes #20098 from gerashegalov/gera/register-SHS-port-conf.

(cherry picked from commit ea956833017fcbd8ed2288368bfa2e417a2251c5)
Signed-off-by: Marcelo Vanzin <vanzin@cloudera.com>

maropu committed:
[SPARK-22937][SQL] SQL elt output binary for binary inputs

## What changes were proposed in this pull request?
This pr modified `elt` to output binary for binary inputs.
`elt` in the current master always output data as a string. But, in some databases (e.g., MySQL), if all inputs are binary, `elt` also outputs binary (Also, this might be a small surprise).
This pr is related to #19977.

## How was this patch tested?
Added tests in `SQLQueryTestSuite` and `TypeCoercionSuite`.

Author: Takeshi Yamamuro <yamamuro@apache.org>

Closes #20135 from maropu/SPARK-22937.

(cherry picked from commit e8af7e8aeca15a6107248f358d9514521ffdc6d3)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

liyinan926 committed:
[SPARK-22960][K8S] Revert use of ARG base_image in images

## What changes were proposed in this pull request?

This PR reverts the `ARG base_image` before `FROM` in the images of driver, executor, and init-container, introduced in https://github.com/apache/spark/pull/20154. The reason is Docker versions before 17.06 do not support this use (`ARG` before `FROM`).

## How was this patch tested?

Tested manually.

vanzin foxish kimoonkim

Author: Yinan Li <liyinan926@gmail.com>

Closes #20170 from liyinan926/master.

(cherry picked from commit bf65cd3cda46d5480bfcd13110975c46ca631972)
Signed-off-by: Marcelo Vanzin <vanzin@cloudera.com>

icexelloss committed:
[SPARK-22930][PYTHON][SQL] Improve the description of Vectorized UDFs for non-deterministic cases

## What changes were proposed in this pull request?

Add tests for using non deterministic UDFs in aggregate.

Update pandas_udf docstring w.r.t to determinism.

## How was this patch tested?
test_nondeterministic_udf_in_aggregate

Author: Li Jin <ice.xelloss@gmail.com>

Closes #20142 from icexelloss/SPARK-22930-pandas-udf-deterministic.

(cherry picked from commit f2dd8b923759e8771b0e5f59bfa7ae4ad7e6a339)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

Commit:
[SPARK-22793][SQL] Memory leak in Spark Thrift Server

# What changes were proposed in this pull request?
1. Start HiveThriftServer2.
2. Connect to thriftserver through beeline.
3. Close the beeline.
4. repeat step2 and step 3 for many times.
we found there are many directories never be dropped under the path `hive.exec.local.scratchdir` and `hive.exec.scratchdir`, as we know the scratchdir has been added to deleteOnExit when it be created. So it means that the cache size of FileSystem `deleteOnExit` will keep increasing until JVM terminated.

In addition, we use `jmap -histo:live [PID]`
to printout the size of objects in HiveThriftServer2 Process, we can find the object `org.apache.spark.sql.hive.client.HiveClientImpl` and `org.apache.hadoop.hive.ql.session.SessionState` keep increasing even though we closed all the beeline connections, which may caused the leak of Memory.

# How was this patch tested?
manual tests

This PR follw-up the https://github.com/apache/spark/pull/19989

Author: zuotingbing <zuo.tingbing9@zte.com.cn>

Closes #20029 from zuotingbing/SPARK-22793.

(cherry picked from commit be9a804f2ef77a5044d3da7d9374976daf59fc16)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

fjh100456 committed:
[SPARK-21786][SQL] When acquiring 'compressionCodecClassName' in 'ParquetOptions', `parquet.compression` needs to be considered.

[SPARK-21786][SQL] When acquiring 'compressionCodecClassName' in 'ParquetOptions', `parquet.compression` needs to be considered.

## What changes were proposed in this pull request?
Since Hive 1.1, Hive allows users to set parquet compression codec via table-level properties parquet.compression. See the JIRA: https://issues.apache.org/jira/browse/HIVE-7858 . We do support orc.compression for ORC. Thus, for external users, it is more straightforward to support both. See the stackflow question: https://stackoverflow.com/questions/36941122/spark-sql-ignores-parquet-compression-propertie-specified-in-tblproperties
In Spark side, our table-level compression conf compression was added by #11464 since Spark 2.0.
We need to support both table-level conf. Users might also use session-level conf spark.sql.parquet.compression.codec. The priority rule will be like
If other compression codec configuration was found through hive or parquet, the precedence would be compression, parquet.compression, spark.sql.parquet.compression.codec. Acceptable values include: none, uncompressed, snappy, gzip, lzo.
The rule for Parquet is consistent with the ORC after the change.

Changes:
1.Increased acquiring 'compressionCodecClassName' from `parquet.compression`,and the precedence order is `compression`,`parquet.compression`,`spark.sql.parquet.compression.codec`, just like what we do in `OrcOptions`.

2.Change `spark.sql.parquet.compression.codec` to support "none".Actually in `ParquetOptions`,we do support "none" as equivalent to "uncompressed", but it does not allowed to configured to "none".

3.Change `compressionCode` to `compressionCodecClassName`.

## How was this patch tested?
Add test.

Author: fjh100456 <fu.jinhua6@zte.com.cn>

Closes #20076 from fjh100456/ParquetOptionIssue.

(cherry picked from commit 7b78041423b6ee330def2336dfd1ff9ae8469c59)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

HyukjinKwon committed:
[SPARK-22901][PYTHON][FOLLOWUP] Adds the doc for asNondeterministic for wrapped UDF function

## What changes were proposed in this pull request?

This PR wraps the `asNondeterministic` attribute in the wrapped UDF function to set the docstring properly.

```python
from pyspark.sql.functions import udf
help(udf(lambda x: x).asNondeterministic)
```

Before:

```
Help on function <lambda> in module pyspark.sql.udf:

<lambda> lambda
(END
```

After:

```
Help on function asNondeterministic in module pyspark.sql.udf:

asNondeterministic()
    Updates UserDefinedFunction to nondeterministic.

    .. versionadded:: 2.3
(END)
```

## How was this patch tested?

Manually tested and a simple test was added.

Author: hyukjinkwon <gurwls223@gmail.com>

Closes #20173 from HyukjinKwon/SPARK-22901-followup.

gatorsmile committed:
[HOTFIX] Fix style checking failure

## What changes were proposed in this pull request?
This PR is to fix the  style checking failure.

## How was this patch tested?
N/A

Author: gatorsmile <gatorsmile@gmail.com>

Closes #20175 from gatorsmile/stylefix.

(cherry picked from commit 9a7048b2889bd0fd66e68a0ce3e07e466315a051)
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

maropu committed:
[SPARK-22973][SQL] Fix incorrect results of Casting Map to String

## What changes were proposed in this pull request?
This pr fixed the issue when casting maps into strings;
```
scala> Seq(Map(1 -> "a", 2 -> "b")).toDF("a").write.saveAsTable("t")
scala> sql("SELECT cast(a as String) FROM t").show(false)
+----------------------------------------------------------------+
|a                                                               |
+----------------------------------------------------------------+
|org.apache.spark.sql.catalyst.expressions.UnsafeMapData38bdd75d|
+----------------------------------------------------------------+
```
This pr modified the result into;
```
+----------------+
|a               |
+----------------+
|[1 -> a, 2 -> b]|
+----------------+
```

## How was this patch tested?
Added tests in `CastSuite`.

Author: Takeshi Yamamuro <yamamuro@apache.org>

Closes #20166 from maropu/SPARK-22973.

(cherry picked from commit 18e94149992618a2b4e6f0fd3b3f4594e1745224)
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

fjh100456 committed:
Merge pull request #1 from apache/master

Merge from apache/spark

fjh100456 committed:
Merge remote-tracking branch 'upstream/branch-2.3'

fjh100456 committed:
Merge pull request #2 from apache/master

Update to upstream master

SparkQA commented:
**[Test build #85780 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85780/testReport)** for PR 20087 at commit [`584cdc2`](https://github.com/apache/spark/commit/584cdc250e1a3a32a1666da4aeb69d5ecb60902b).

SparkQA commented:
**[Test build #85780 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85780/testReport)** for PR 20087 at commit [`584cdc2`](https://github.com/apache/spark/commit/584cdc250e1a3a32a1666da4aeb69d5ecb60902b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85780/
Test FAILed.

fjh100456 committed:
Fix test issue

Fix test issue

SparkQA commented:
**[Test build #85787 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85787/testReport)** for PR 20087 at commit [`5b150bc`](https://github.com/apache/spark/commit/5b150bc643e301ec063704303aebeea87de02e83).

SparkQA commented:
**[Test build #85787 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85787/testReport)** for PR 20087 at commit [`5b150bc`](https://github.com/apache/spark/commit/5b150bc643e301ec063704303aebeea87de02e83).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85787/
Test FAILed.

fjh100456 commented:
Retest please. @gatorsmile

fjh100456 committed:
Merge branch 'master' of https://github.com/fjh100456/spark

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #85820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85820/testReport)** for PR 20087 at commit [`5b150bc`](https://github.com/apache/spark/commit/5b150bc643e301ec063704303aebeea87de02e83).

SparkQA commented:
**[Test build #85820 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85820/testReport)** for PR 20087 at commit [`5b150bc`](https://github.com/apache/spark/commit/5b150bc643e301ec063704303aebeea87de02e83).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85820/
Test PASSed.

fjh100456 committed:
consider the precedence of `hive.exec.compress.output`

SparkQA commented:
**[Test build #85944 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85944/testReport)** for PR 20087 at commit [`4b89b44`](https://github.com/apache/spark/commit/4b89b44b2b4a9a1ffee01da13a3e55cdd3aa260d).

fjh100456 commented:
@gatorsmile 
I'd change the precedence. 
For Parquet, if `hive.exec.compress.output` is true, keep the old precedence, otherwise, get compression from HiveOption.
For Orc, because `hive.exec.compress.output` always take no effect, we get compression from HiveOption directly.

If this is ok, I'll change the test case. Any suggestion?

SparkQA commented:
**[Test build #85944 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85944/testReport)** for PR 20087 at commit [`4b89b44`](https://github.com/apache/spark/commit/4b89b44b2b4a9a1ffee01da13a3e55cdd3aa260d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85944/
Test PASSed.

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+    }
+  }
+
+  private val maxRecordNum = 500
```
Reduce it to 50 for decreasing the execution time

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+        block <- footer.getParquetMetadata.getBlocks.asScala
+        column <- block.getColumns.asScala
+      } yield column.getCodec.name()
+      case "orc" => new File(path).listFiles().filter{ file =>
```
Nit: add a space before `{`

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+      tableName: String,
+      partition: Option[String]): Unit = {
+    val partitionInsert = partition.map(p => s"partition (p='$p')").mkString
+    sql(
```
This is INSERT after CREATE TABLE. We also need to test/fix another common cases, CTAS [CREATE TABLE AS SELECT]

fjh100456 replied:
CTAS statement is not allowed to create a partitioned table using Hive's file formats. So I use the syntax of `CREATE TABLE tableName USING ... OPTIONS (...) PARTITIONED BY...` to create a table.  

However, it seems to be different from non-partitioned hive table when convertMetastore is true.For non-partitioned hive table, session-level will take effect, but for table created by CTAS, table-level takes effect.

And if I merge the code of your PR(#20120), they would be consistent, table-level compression will take effect.

Should I fix it after your PR closed?

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+      withTempDir { tmpDir =>
+        withTable(tableName) {
+          createTable(tmpDir, tableName, isPartitioned, format, Option(codecName))
+          val partition = if (isPartitioned) Some("test") else None
```
`partition ` -> `partitionValue`

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+    withTempDir { tmpDir =>
+      withTable(tableName) {
+        createTable(tmpDir, tableName, isPartitioned, format, compressionCodec)
+        val partition = if (isPartitioned) Some("test") else None
```
`partition ` -> `partitionValue`

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+      tableCompressionCodecs: List[String])
+      (assertionCompressionCodec: (Option[String], String, String, Long) => Unit): Unit = {
+    withSQLConf(getConvertMetastoreConfName(format) -> convertMetastore.toString) {
+      tableCompressionCodecs.foreach { tableCompression =>
```
You are setting compressionCodecs and tableCompressionCodecs to the same one. I think we can set it to different ones to know which one is being used.

fjh100456 replied:
Shouldn't we test the precedence of compressionCodecs in effect? Here I just want to verify the precedence.

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+            val compression = Option(tableCompression)
+            checkCompressionCodecForTable(format, isPartitioned, compression) {
+              case (realCompressionCodec, tableSize) => assertionCompressionCodec(compression,
+                sessionCompressionCodec, realCompressionCodec, tableSize)
```
```
case (realCompressionCodec, tableSize) =>
  assertionCompressionCodec(
    compression, sessionCompressionCodec, realCompressionCodec, tableSize)
```

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+          isPartitioned,
+          convertMetastore,
+          compressionCodecs = compressCodecs,
+          tableCompressionCodecs = compressCodecs) {
```
Also add another scenario. 
```
compressionCodecs = Nil,
tableCompressionCodecs = compressCodecs 
```

fjh100456 replied:
Do you mean only with table-level compression? Actually, `compressionCodecs` is the session-level compressionCodec, even if set to `Nil` or `null` here, it still takes the default value `snappy`.  
In the first test case ``` test("both table-level and session-level compression are set") ```, isn't it already contained?

gatorsmile replied:
OK.

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+          isPartitioned,
+          convertMetastore,
+          compressionCodecs = compressCodecs,
+          tableCompressionCodecs = List(null)) {
```
`List(null)` -> `Nil`

fjh100456 replied:
If change to `Nil`, the follow function may requires special handling of this situation. Set to `null` is to get a `None`.  Can we keep it?

``` scala
private def checkTableCompressionCodecForCodecs(
      format: String,
      isPartitioned: Boolean,
      convertMetastore: Boolean,
      compressionCodecs: List[String],
      tableCompressionCodecs: List[String])
      (assertionCompressionCodec: (Option[String], String, String, Long) => Unit): Unit = {
    withSQLConf(getConvertMetastoreConfName(format) -> convertMetastore.toString) {
      tableCompressionCodecs.foreach { tableCompression =>
        compressionCodecs.foreach { sessionCompressionCodec =>
          withSQLConf(getSparkCompressionConfName(format) -> sessionCompressionCodec) {
            // 'tableCompression = null' means no table-level compression
            val compression = Option(tableCompression)
            checkCompressionCodecForTable(format, isPartitioned, compression) {
              case (realCompressionCodec, tableSize) =>
                assertionCompressionCodec(
                  compression, sessionCompressionCodec, realCompressionCodec, tableSize)
            }
          }
        }
      }
    }
  }
```

gatorsmile replied:
It does not work?

fjh100456 replied:
If set to `Nil`,  ` foreach` statement will not go in and the testcase will do nothing. So a special handling must be done. Set to `null` will get a `None` value by ```val compression = Option(tableCompression)``` and it is exactly what I want.

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+            val relCompressionCodecs =
+              if (isPartitioned) compressCodecs.flatMap { codec =>
+                getTableCompressionCodec(s"$tablePath/p=$codec", format)
+              } else getTableCompressionCodec(tablePath, format)
```
```
} else {
  getTableCompressionCodec(tablePath, format)
}
```

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,321 @@
+          createTable(tmpDir, tableName, isPartitioned, format, None)
+          withTable(tableName) {
+            compressCodecs.foreach { compressionCodec =>
+              val partition = if (isPartitioned) Some(compressionCodec) else None
```
`partition` -> `partitionValue`

AmplabJenkins commented:
Can one of the admins verify this patch?

fjh100456 committed:
Resume to private and add public function

fjh100456 committed:
Resume to private and add public function

fjh100456 committed:
Fix test issue

SparkQA commented:
**[Test build #86383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86383/testReport)** for PR 20087 at commit [`99271d6`](https://github.com/apache/spark/commit/99271d670a0aed444ad624d56304d94490eed0cb).

fjh100456 commented:
@gatorsmile 
I had fix the test case for CTAS, but it may not pass the test, until merge the code of your PR #20120

SparkQA commented:
**[Test build #86383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86383/testReport)** for PR 20087 at commit [`99271d6`](https://github.com/apache/spark/commit/99271d670a0aed444ad624d56304d94490eed0cb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86383/
Test FAILed.

gatorsmile commented:
@fjh100456 Thanks for working on it! It is pretty close to be merged.

fjh100456 committed:
Fix test issue

fjh100456 committed:
Fix style issue

SparkQA commented:
**[Test build #86409 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86409/testReport)** for PR 20087 at commit [`5b5e1df`](https://github.com/apache/spark/commit/5b5e1df983af6ff03ec6ef6c83208c8b25af93e2).

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala:
```diff
@@ -0,0 +1,354 @@
+            compressionCodecs = compressCodecs,
+            tableCompressionCodecs = List(null)) {
+            case
+              (tableCodec, sessionCodec, realCodec, tableSize) =>
```
Nit: the style issue.

fjh100456 replied:
Oops, I made a mistake. Thank you !

SparkQA commented:
**[Test build #86409 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86409/testReport)** for PR 20087 at commit [`5b5e1df`](https://github.com/apache/spark/commit/5b5e1df983af6ff03ec6ef6c83208c8b25af93e2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86409/
Test FAILed.

fjh100456 committed:
Fix style issue

SparkQA commented:
**[Test build #86411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86411/testReport)** for PR 20087 at commit [`118f788`](https://github.com/apache/spark/commit/118f7880bdcf26ba7394a2cc7fac2e0eae707d6f).

SparkQA commented:
**[Test build #86411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86411/testReport)** for PR 20087 at commit [`118f788`](https://github.com/apache/spark/commit/118f7880bdcf26ba7394a2cc7fac2e0eae707d6f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86411/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #86415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86415/testReport)** for PR 20087 at commit [`118f788`](https://github.com/apache/spark/commit/118f7880bdcf26ba7394a2cc7fac2e0eae707d6f).

SparkQA commented:
**[Test build #86415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86415/testReport)** for PR 20087 at commit [`118f788`](https://github.com/apache/spark/commit/118f7880bdcf26ba7394a2cc7fac2e0eae707d6f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86415/
Test PASSed.

gatorsmile commented:
LGTM

Thanks! Merged to master/2.3

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala:
```diff
@@ -55,18 +55,28 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
+          // For ORC,"mapreduce.output.fileoutputformat.compress",
+          // "mapreduce.output.fileoutputformat.compress.codec", and
+          // "mapreduce.output.fileoutputformat.compress.type"
+          // have no impact because it uses table properties to store compression information.
```
Although this is the existing behavior, but could you investigate how Hive behaves when `Parquet.Compress`  is set. https://issues.apache.org/jira/browse/HIVE-7858 Is it the same as ORC?

fjh100456 replied:
Surely, I'll do it this days.

fjh100456 replied:
For parquet, using a hive client, `parquet.compression` has a higher priority than  `mapreduce.output.fileoutputformat.compress`. And table-level compression( set by tblproperties) has the highest priority.  `parquet.compression` set by cli also has a higher priority than `mapreduce.output.fileoutputformat.compress`.

After this pr, the priority does not changed. If table-level compression was set, other compression would not take effect, even though `mapreduce.output....` were set, which is the same with hive. But `parquet.compression` set by spark cli does not take effect, unless set `hive.exec.compress.output` to true. This may because we do not get `parquet.compression` from the session, and I wonder if it's necessary because we have `spark.sql.parquet.comression` instead.

For orc, `hive.exec.compress.output` and `mapreduce.output....` have no impact really, but table-leval  compression (set by tblproperties) always take effect.  `orc.compression` set by spark cli does not take effect too, even though  set `hive.exec.compress.output` to true, which is differet with parquet. 
Another question, the comment say `it uses table properties to store compression information`, actully, by manul test, I found orc-tables also can have mixed compressions, and the data can be read together correctly.

My Hive version for this test is 1.1.0.  Actully it's a little difficut for me to get a higher version runable Hive client.

gatorsmile replied:
The comment might not be correct now. We need to follow what the latest Hive works, if possible. The best way to try Hive (and the other RDBMS) is using docker. Maybe you can try the docker?

fjh100456 replied:
Ok, I'll try it.
